File: src/main/java/com/github/cao/awa/kalmia/message/DeletedMessage.java
package com.github.cao.awa.kalmia.message;

import com.github.cao.awa.apricot.io.bytes.reader.BytesReader;
import com.github.cao.awa.kalmia.mathematic.base.SkippedBase256;
import com.github.cao.awa.kalmia.message.digest.DigestData;
import com.github.cao.awa.viburnum.util.bytes.BytesUtil;

public class DeletedMessage extends Message {
    private final long sender;
    private final DigestData digestData;

    public DeletedMessage(long sender, DigestData digestData) {
        this.sender = sender;
        this.digestData = digestData;
    }

    public long getSender() {
        return this.sender;
    }

    @Override
    public byte[] toBytes() {
        return BytesUtil.concat(new byte[]{- 1},
                                SkippedBase256.longToBuf(this.sender),
                                this.digestData.toBytes()
        );
    }

    public static DeletedMessage create(BytesReader reader) {
        if (reader.read() == - 1) {
            long sender = SkippedBase256.readLong(reader);

            DigestData digestData = DigestData.create(reader);

            return new DeletedMessage(sender,
                                      digestData
            );
        } else {
            return null;
        }
    }

    @Override
    public DigestData digest() {
        return this.digestData;
    }
}

File: src/main/java/com/github/cao/awa/kalmia/message/deleted/DeletedMessage.kt
package com.github.cao.awa.kalmia.message.deleted

import com.github.cao.awa.apricot.io.bytes.reader.BytesReader
import com.github.cao.awa.kalmia.identity.LongAndExtraIdentity
import com.github.cao.awa.kalmia.message.Message
import com.github.cao.awa.kalmia.message.digest.DigestData
import com.github.cao.awa.kalmia.message.display.ClientMessageContent
import com.github.cao.awa.viburnum.util.bytes.BytesUtil

class DeletedMessage : Message {
    companion object {
        private val HEADER = byteArrayOf(-1)

        @JvmStatic
        fun create(reader: BytesReader): DeletedMessage? {
            return if (reader.read().toInt() == -1) {
                val identity = LongAndExtraIdentity.read(reader)
                val sender = LongAndExtraIdentity.read(reader)
                val digestData = DigestData.create(reader)

                DeletedMessage(
                    identity,
                    sender,
                    digestData
                )
            } else {
                null
            }
        }
    }

    private val sender: LongAndExtraIdentity
    private val digestData: DigestData

    constructor(sender: LongAndExtraIdentity, digestData: DigestData) {
        this.sender = sender
        this.digestData = digestData
    }

    constructor(
        identity: LongAndExtraIdentity,
        sender: LongAndExtraIdentity,
        digestData: DigestData
    ) : super(identity) {
        this.sender = sender
        this.digestData = digestData
    }

    override fun sender(): LongAndExtraIdentity = this.sender

    override fun header(): ByteArray = HEADER

    override fun display(): ClientMessageContent {
        return ClientMessageContent(
            sender(),
            "DeletedMessage{sender=${sender()}, digest=${digest().value36()}}",
            """
                        This message has been deleted
                        sender is: ${sender()}
                      digest is: ${digest().value36()}"""
                .trimIndent()
        )
    }

    override fun toBytes(): ByteArray {
        return BytesUtil.concat(
            header(),
            identity().toBytes(),
            sender().toBytes(),
            digest().serialize()
        )
    }

    override fun digest(): DigestData = this.digestData
}